### DISCUSSION TITLE
Lambda expressioned pattern matching

### DISCUSSION THREAD

**Discussion by eh-robp3** [dotnet/csharplang#8897] [Language Ideas]
In my opinion C# language is missing some pattern matching features at least most important is lambda expression in matching clause.
This is strongly inspired by groovy https://blog.mrhaki.com/2009/08/groovy-goodness-switch-statement.html
`case { it instanceof Integer && it < 50 }`

Having this feature build in language I could imagine that all kind of problems would be easly solved (regex, string comparisions, more complex scenarios)
Example of usage (which doesn't require keyword `when` afterwards pattern) :
```
List<string> list = new List<string>(){"1","2","3","4","5"};
 
var result = list switch
{
    [{} first { it => it.StartsWith("a", StringComparison.OrdinalIgnoreCase) },..] => "first element starts with 'a'",
    [{} first { it => Regex.IsMatch(it, "[0-9]+", RegexOptions.IgnoreCase) },..] => "first element is a number",
    _ => string.Empty
};
```
Please shere your feedback on this feature and vote for more detailed discussion.

**Comment by KennethHoff** [comment]
There is something like it already, but it requires moves it after a `when` clause.

```diff
-    [{} first { it => it.StartsWith("a", StringComparison.OrdinalIgnoreCase) },..] => "first element starts with 'a'",
+    [{} first, ..] when first.StartsWith("a", StringComparison.OrdinalIgnoreCase) => "first element starts with 'a'",
```

**Reply by eh-robp3** [reply]
Yes, I'm fully awere of this and I've done like that on my code. I would like to propose solution to avoid `when` keyword and thus implement logic inside pattern matching region. This could be a nice extension of existing types of patter matchings:
`[{Length: > 10 and {it => it.Contains("a")}}, ..] => "first element is longer than 10 and contains "a"",`

**Reply by CyrusNajmabadi** [reply]
>  I would like to propose solution to avoid when keyword and thus implement logic inside pattern matching region

Why is this necessary?  Is it just about rearranging the existing code that can be written?

**Reply by HaloFour** [reply]
It feels like the ask is to allow `switch` `case` arms to evaluate arbitrary Boolean expressions.  That is at least how the feature works in Groovy.

Or maybe it's about allowing for case guard clauses to be declared within a recursive pattern match expression rather than having to be declared at the end.

**Reply by eh-robp3** [reply]
Yop. I would like to have `switch case` to act the same way as in groovy (i think there are more languages with support lambda expressions [?]).

**Reply by HaloFour** [reply]
I'm not aware of any other languages that use lambda syntax in pattern matching like Groovy.  Kotlin has case guards in lieu of pattern matching, but that's probably as close as it gets and it's quite limited.

I personally don't like the use of lambda syntax here.  You're not declaring a function, just a Boolean expression.  To that end, if anything were to happen here, I'd rather explore locality of case guards, or possibly even a degenerate form if the guarded pattern happened to be only a variable pattern.

```cs
var result = list switch
{
    [{} first when first.StartsWith("a", StringComparison.OrdinalIgnoreCase),..] => "first element starts with 'a'",
    [{} first when Regex.IsMatch(first, "[0-9]+", RegexOptions.IgnoreCase),..] => "first element is a number",
    _ => string.Empty
};
```

**Comment by 333fred** [comment]
See https://github.com/dotnet/csharplang/issues/8613, which covers this type of case.

**Reply by eh-robp3** [reply]
This seems just a half of solution. In [#8613](https://github.com/dotnet/csharplang/issues/8613) do the job but it requires some code in class which will be later used in pattern matching. 

```
public static bool IsInteger(this string str) => int.TryParse(str, out _);

if (myStr is Integer()) ...
```
I just would like to express my intention on labda pattern matching (`string item {it => my lambda expression based on it }`. 

```
var result = o switch
{
    string s when int.Parse(s)  => 0,
    string item { it => it.TryParse(out var r) && it.IsOdd() } => r,
    _ => 1
};
```

**Reply by HaloFour** [reply]
If user-defined (or active) patterns also permit inputs, they should permit the following:

```cs
List<string> list = new List<string>(){"1","2","3","4","5"};
 
var result = list switch
{
    [StartsWith("a", StringComparison.OrdinalIgnoreCase), ..] => "first element starts with 'a'",
    [Regex("[0-9]+", RegexOptions.IgnoreCase), ..] => "first element is a number",
    _ => string.Empty
};
```

**Comment by eh-robp3** [comment]
Thanks for discussion. I will close our conversation.
